### ISSUE TITLE
🚧 | Y3: Efficient subnet consensus protocols

### ISSUE THREAD

**Issue by jsoares** [protocol/ConsensusLab#9]
## Description

Our goal is to design and implement scalable, efficient consensus protocols for subnets (i.e. anything below the top-level Filecoin consensus). This should enable secure, low-latency operation up to ~500 nodes per subnet.

These protocols will be integrated with [Eudico](https://github.com/filecoin-project/eudico), the research clone of the [Lotus](https://github.com/filecoin-project/lotus) Filecoin client, supporting a hierarchy of subnets, each running its own instance of a consensus protocol.

To achieve our, we develop [Mir](https://github.com/filecoin-project/mir), a framework for implementing distributed protocols. The first protocol to be implemented is [ISS](https://github.com/filecoin-project/mir/tree/main/pkg/iss), a multi-leader BFT-style consensus protocol, in conjunction with [Narwhal](https://arxiv.org/abs/2105.11827), a state-of-the-art mempool implementation enabling sale-out throughput.

The design should be, however, general enough, and its implementation sufficiently modular, to make it easy to implement various different consensus protocols and select any of them for any particular subnet deployment.

## Scope
- Specify mid-tear and leaf-level consensus
- Implement a modular and scalable (up to low hundred of nodes) consensus solution
- Integrate the solution with the the (experimental) Filecoin client Eudico.

## Resources
- [Mir](https://github.com/filecoin-project/mir) distributed protocol implementation framework
- [ISS](https://github.com/filecoin-project/mir/tree/main/pkg/iss) consensus protocol implementation (in progress)
- [Incremental Design of Eudico's Ordering Layer](https://hackmd.io/1vxyguGMSlWZM9_s8PCFbw?view) (describes high-level ideas and logical abstractions)
- [Implementation of Eudico’s Ordering Layer](https://hackmd.io/P59lk4hnSBKN5ki5OblSFg)
- [Testing and verification plan](https://hackmd.io/OEIctHWCQV6tmJwVUGNX0Q) (By @dnkolegov , WIP)

## Papers
- [ISS protocol](https://dl.acm.org/doi/abs/10.1145/3492321.3519579)
- [Narwhal and Tusk](https://arxiv.org/abs/2105.11827)
- [Jolteon and Ditto](https://arxiv.org/abs/2106.10362)
- [Bullshark](https://arxiv.org/abs/2201.05677)

## Talks
soon™️  

#### 2022-04-07, MirBFT 
[![Watch the video](https://img.youtube.com/vi/_dwGXxa8hS0/hqdefault.jpg)](https://youtu.be/_dwGXxa8hS0)

#### 2022-06-02, Crash failures in Mir and its integration with Eudico
[![Watch the video](https://img.youtube.com/vi/1bOXCb29Pmw/hqdefault.jpg)](https://youtu.be/1bOXCb29Pmw)

#### 2022-08-12, Reproducible Integration Testing in Mir
[![Watch the video](https://img.youtube.com/vi/1-08R2tGHtc/hqdefault.jpg)](https://youtu.be/1-08R2tGHtc)

#### 2022-08-12, Taking Pseudocode To An Implementation With Mir Framework
[![Watch the video](https://img.youtube.com/vi/ENDC2IqjzRU/hqdefault.jpg)](https://youtu.be/ENDC2IqjzRU)

#### 2022-09-01, Reconfigurable SMR with Mir 
[![Watch the video](https://img.youtube.com/vi/4WIyw1MvzkE/hqdefault.jpg)](https://youtu.be/4WIyw1MvzkE)

#### 2022-09-01, Dynamically adding new nodes to Eudico with Mir 
[![Watch the video](https://img.youtube.com/vi/1006G2OfGt8/hqdefault.jpg)](https://youtu.be/1006G2OfGt8)

**Comment by matejpavlovic** [comment]
Notes from meeting of @matejpavlovic , @dnkolegov , @adlrocha : https://docs.google.com/document/d/1FqmaN2rVscfvRdtXj7G_nOwA-EVP1Rc9lCGOcz-NvVs/edit?usp=sharing

**Comment by jsoares** [comment]
### 2022-02-21 meeting notes

* Discussion with Alfonso and Denis around implementation approach. Synchronised on the state of each part of the project. Tendermint almost done. Next upcoming chunk is on the Eudico side (~2 weeks). In parallel, we’ll look into implementing MirBFT stub. 
* We’ll also look into how ISS works and identify the key parts of the ISS protocol that need to be implemented to have a working prototype.
    * Marko: We should come up with roadmap for this
        * Timeline: With testnet in Q4, what can we expect to have ready in time? 
            * Having stable interface is the priority.
            * Estimating effort and timeline is next step — aiming to have someone by Monday. Let’s try for monthly milestones and zero-mean error. 
* Narwhal: first focus on the framework, interface and requirements for usage. For now, plan to use default mempool. Alfonso will be supporting on Eudico side. Immediate AIs: share ISS paper with Denis and go through MirBFT.
    * Is it clear that Narwahl can be added on top of Mir-BFT? Is there a unifying architecture?
        * Narwhal and MirBFT are not interchangeable. Narwhal can use MirBFT as a blackbox ordering service. The architecture should be modular, with MirBFT and Narwhal as specific instances.
        * We may lose some specific properties/advantages in favour of generality.
* How should we do state transfer work in Eudico? 
    * Have a snapshot of the state tree. Unclear how efficient it will be. This is synchronised to a specific epoch and consistent across nodes. We can also add reconfiguration messages — these can be part of a block and be part of the total order.

**Comment by jsoares** [comment]
## 2000-02-28

### ✋ Attendees
- @matejpavlovic 
- @dnkolegov 
- @adlrocha 
- @jsoares 
- @LefKok 
- @vukolic 

### 📣 Updates
- Discussing architecture and making a list of to-do items to make MirBFT a consensus protocol for Eudico
- First milestone define
    - M1: minimal demoable implementation
        - Eudico client-service interface (~ that for Tendermint) [Denis, 3 wks]
        - Implementation of PBFT subprotocol in ISS [Matej, 3 wks]
        - State garbage collection and transfer [4 wks]
        - Prepare demo [1 day]

### 🧵 Discussion 
- Did we consider alternatives to PBFT?
    - Alternatives considered but PBFT is the most settled and general. Fastest to implement. Good performance in previous experiments.
    - We can try others in parallel but preference for the team to sart there. 
- Timeline for the overall project
    - End of Q3 realistic for testnet
    - Only critical items after M1 is 
        - Reconfiguration
        - Persistent request store
        - Persistent write-ahead log (trickier)
    - Let's try to organise the follow-up milestones
- Matej + Denis + Sergey sounds like full staffing; shouldn't need any more capacity. Some parts are not highly parallelisable.
- We may be able to reuse testing framework from Lotus/Eudico instead of spending a lot of dev cycles on it

### 🎯 Up next
- Define timeline for M1 (start and end date)
- Define high-level milestones for the rest of the project
- Organise meeting on testing tools and invite Marko

**Comment by jsoares** [comment]
## 2022-03-07

### ✋ Attendees
- @adlrocha 
- @dnkolegov 
- @jsoares 
- @matejpavlovic
- @vukolic 

### 📣 Updates
- Refined [Project Milestones](https://hackmd.io/RYFd1md4QK2JrH8l9Z-bYw?both)
    - Assigned tasks by priority to different milestones
    - Separated parallelizable best-effort tasks
- Added estimated timeline for Milestone #1 (P0)
- Wrote [internship proposal](https://hackmd.io/@matejpavlovic/B1PjywQZc)

### 🧵 Discussion 
- Applicability of VDFs to PBFT vs. other protocols
- Internship proposal comments?
- Robust leader election
    - Want to verify an idea/assumption that PBFT could easily be made robust in this sense. 

### 🎯 Up next
- Continue implementing PBFT sub-protocol
- Check alignment of interests with Srivatsan
- Potentially submit Hyperledger mentorship proposal (deadline on Wed)
- @jsoares to add milestones to roadmap

**Comment by jsoares** [comment]
## 2022-03-21

### ✋ Attendees
- @adlrocha 
- @dnkolegov 
- @jsoares 
- @LefKok 
- @matejpavlovic 
- @vukolic 

### 📣 Updates
- Started implementing PBFT view change
- Consolidated big part of the documentation for MirBFT
- Updated some related [issues on ConsensusLab GitHub](https://github.com/protocol/consensuslab/issues/9), added a few sub-issues

### 🧵 Discussion 
- New target date for MirBFT-Eudico interface 3 weeks into April: OK with Denis?
    - Denis: Yes

### 🎯 Up next
- Continue working on PBFT view change implementation
- Maybe demo (if time allows)

### 🛑 Private
- What went great?
- What can be improved?
    - Potential contributor didn't write back yet
- Discussions
- FYIs

**Comment by jsoares** [comment]
## 2022-03-28

### ✋ Attendees
- @dnkolegov 
- @jsoares 
- @matejpavlovic 
- @vukolic 

### 📣 Updates
- Suggested project for Sergey: State transfer in MirBFT.
- More code on PBFT view change.

### 🧵 Discussion 
- Make [#90 ISS with chain quality ](https://github.com/protocol/ConsensusLab/issues/90) a discussion?
    - It's now a discussion -- let's document today's conversation and other thoughts as they come along.
- Usage of VRFs for bucket assignment.
- Value proposition of Narwhal vs MirBFT
    - And how they go hand-in-hand
    - Should we have a sprint to work out the best direction? Start async discussion; highest priority.
        - Denis can still work on MirBFT integration after itests; meeting on Wednesday
    - Task protocol also at top of mind; simple, co-designed with Narwhal, something to know about when having this discussion
        - Ideas from Hedera/hashgraph?

### 🎯 Up next
- PBFT View change.
- MirBFT usage demo for 7 April.

**Comment by adlrocha** [comment]
From complete ignorance, @matejpavlovic, how does Mir-BFT relates to Node.js? (I am just wary of this kind of generic names :smile: )

**Comment by dnkolegov** [comment]
Why Node? Node-go or go-node says that "Node" is implemented in Go, but what Node is, and how does it relate to BFT?

If we are talking about names:
-  uBFT - Universal BFT or any other xBFT, and then ubft-go,...
- orbitox

**Comment by matejpavlovic** [comment]
Proposing the initial draft of the design document for incremental implementation of Eudico's ordering layer. Comments welcome!
https://hackmd.io/@matejpavlovic/ryd53ZP4c

**Comment by sergefdrv** [comment]
> initial draft of the design document for incremental implementation of Eudico's ordering layer

I left some comments

**Comment by matejpavlovic** [comment]
## 2022-04-21

### ✋ Attendees
- @dnkolegov 
- @jsoares
- @matejpavlovic 
- @sergefdrv 
- @vukolic 

### 📣 Updates
- Wrote [design document](https://hackmd.io/1vxyguGMSlWZM9_s8PCFbw?view)
  on the architecture and modularization of Eudico's ordering layer.
- Started documenting Mir architecture decisions with ADRs
- Updated roadmap on CL GitHub

### 🧵 Discussion 
- Separate Slack channel for Mir implementation?
    - #mir-impl or #mir-implementation
    - Public if not polluting the Filecoin Slack space, otherwise private
- We agree on the proposed high-level design
    - This is not the detailed implementation architecture
- We will start by implementing the first simple variant algorithm-wise and add all necessary algorithm-independent features (reconfiguration, restarts, etc.) before evolving the algorithms

### 🎯 Up next
- Continue implementation
- @matejpavlovic to devise a more concrete implementation architecture

**Comment by matejpavlovic** [comment]
## 2022-05-02

### ✋ Attendees
- @adlrocha 
- @dnkolegov 
- @LefKok 
- @matejpavlovic 
- @sergefdrv 
- @vukolic 

### 📣 Updates
- @matejpavlovic:
    - [Architecture of consensus layer implementation](https://hackmd.io/P59lk4hnSBKN5ki5OblSFg)
    - [ADR proposal](https://hackmd.io/1v_YcDPNS7K0qyo5J6rBkw) for time abstraction in Mir
    - Mir improvements:
      [fix race in test deployment](https://github.com/filecoin-project/mir/issues/20),
      [fix GitHub Actions](https://github.com/filecoin-project/mir/pull/30),
      [fix shutdown bug](https://github.com/filecoin-project/mir/issues/36),
      [add linter check as GitHub Action](https://github.com/filecoin-project/mir/pull/42)
    - Still implementing PBFT view change
- @sergefdrv:
    - Resolved [Basic ISS checkpoints](https://github.com/filecoin-project/mir/issues/9) - Create certified checkpoints and produce debug output when they are created
- @dnkolegov:
    - Launching Mir in subnet via rootnet
    - Bug fixing

### 🧵 Discussion 
- @matejpavlovic: Time abstraction
    - Example of timer usage to ADR
- @matejpavlovic: Sergey's launchpad
    - No for the moment
- MV: do we have upcoming (May) milestones defined? How will we evaluate progress against our goals?
    - https://github.com/protocol/ConsensusLab/issues/62
- DK: Eudico miner within Mir: at present we can't know who has proposed (mined) a block to propogate this to Filecoin blocks
    - DK to open an issue in Mir around this
        - https://github.com/filecoin-project/mir/issues/43
    - Explicitly ping Alfonso if we need his input

### 🎯 Up next
- @matejpavlovic:
    - Implement PBFT view change
    - Reach consensus on technical decisions
    - Continue improving Mir
- @sergefdrv:
    - Open discussion on Mir framework architecture concerns
    - Feedback on [Implementation of Eudico’s Ordering Layer](https://hackmd.io/P59lk4hnSBKN5ki5OblSFg?view) and [Time abstraction in Mir](https://hackmd.io/1v_YcDPNS7K0qyo5J6rBkw)
    - Work on [ISS state garbage collection](https://github.com/filecoin-project/mir/issues/10)

**Comment by matejpavlovic** [comment]
## 2022-05-09

### ✋ Attendees
- @adlrocha 
- @dnkolegov 
- ~~@jsoares~~ 
- @LefKok 
- @matejpavlovic 
- @sergefdrv 
- @vukolic 

### 📣 Updates
- @sergefdrv 
    - [Discussion on Mir framework architecture](https://github.com/filecoin-project/mir/discussions/46)
    - [Working](https://github.com/filecoin-project/mir/pull/52) on [ISS state garbage collection](https://github.com/filecoin-project/mir/issues/10)
- @matejpavlovic
    -  [Discussion on Mir framework architecture](https://github.com/filecoin-project/mir/discussions/46)
    -  Improving Mir
    -  Implementing PBFT view change
    -  Internship with Andrei for implementing Narwhal in Mir
- @dnkolegov 
    - I was working on intergration Eudico with Mir
        - bug fixing
        - refactoring
    - Now we can define a miner for Filecoin blocks

### 🧵 Discussion 
- Mir architecture
    - Try using the current event-based architecture and see if we need to work a lot against it.

### 🎯 Up next
- @matejpavlovic
    - Continue implementing PBFT and Mir
    - Interface to Eudico definition
- @sergefdrv 
    - Finish [ISS state garbage collection](https://github.com/filecoin-project/mir/issues/10)
    - Work on [ISS state transfer](https://github.com/filecoin-project/mir/issues/11)

**Comment by matejpavlovic** [comment]
## 2022-05-16

### ✋ Attendees
- @adlrocha 
- @dnkolegov:
- @jsoares 
- @LefKok 
- @matejpavlovic 
- @sergefdrv 
- @vukolic 

### 📣 Updates
- @dnkolegov
    - Eudico demos
    - Eudico and Mir integration
    - Improving Eudico
    - Initial test concept
    - Discussion with Matej on the Mir design and the next task for me
- @matejpavlovic
    - Implemented [PBFT view change](https://github.com/filecoin-project/mir/pull/60) and a [debugger for Mir](https://github.com/filecoin-project/mir/pull/58)
    - Drafted [list of things to implement for ordering layer](https://hackmd.io/9SPHp_SWSHa8zbwXe9fO3Q)
    - Started designing [Eudico's ordering layer interface ](https://hackmd.io/6v00B72USpav6rDUUhqxAw)
- @sergefdrv
    - Implemented [state garbage collection](https://github.com/filecoin-project/mir/pull/52)
    - Fixed [BFT quorum size calculation](https://github.com/filecoin-project/mir/pull/61)

### 🧵 Discussion 
- How will we do state transfer?
    - Plans for week of 16/05: grab serialised state and transfer as it is; trivial app example, might not work for Eudico.
    - We'll have checkpoints; in the long run, the state transfer within Mir will likely be the same thing. Then we can transfer overall state in Eudico.
    - Mir is there to do ordering, not deal with full state payload.
- Y3-M3 plans
    - Will need some input from Alfonso in the next few days
    - Early focus on perf benchmarks
- Demo on 2 Jun?
    - We can demo Denis' previous work + failure
    - Benchmarks not likely

### 🎯 Up next
- @dnkolegov:
    - Multisig collector
- @matejpavlovic
    - Converge on a spec for Eudico's ordering layer interface
    - Specify concrete issues for the Y3-M3 milestone (ordering layer MVP)
    - Present Mir debugger and show how it is used
- @sergefdrv 
    - [Test memory consumption with/without checkpoint garbage collection](https://github.com/filecoin-project/mir/issues/10)
    - [Work on state transfer](https://github.com/filecoin-project/mir/issues/11)

**Comment by jsoares** [comment]
## 2022-05-23

### ✋ Attendees
- @dnkolegov 
- @jsoares 
- @LefKok 
- @matejpavlovic 
- @sergefdrv 
- @vukolic 

### 📣 Updates
- @matejpavlovic 
    - Outlined [todos and next steps](https://hackmd.io/9SPHp_SWSHa8zbwXe9fO3Q)
    - [Ordering protocol interface](https://hackmd.io/6v00B72USpav6rDUUhqxAw) (with the team)
    - Presentation of mircat debugger
    - [Timer module for Mir](https://github.com/filecoin-project/mir/pull/67)
    - Will be busy with SRDS reviews (deadline on Thu)
- @sergefdrv 
    - [Fixed memory leak in ISS](https://github.com/filecoin-project/mir/pull/63)
    - [Added simple memory footprint benchmark in Mir](https://github.com/filecoin-project/mir/pull/64)
    - [Fixed transitioning into new epoch in ISS](https://github.com/filecoin-project/mir/pull/66)
    - Y3 discussions with the team ([this](https://hackmd.io/9SPHp_SWSHa8zbwXe9fO3Q) and [this](https://hackmd.io/6v00B72USpav6rDUUhqxAw))
    - [Implemented state transfer in ISS](https://github.com/filecoin-project/mir/pull/68)
- @dnkolegov
    - Discussed new [Eudico Ordering Layer Interface](https://hackmd.io/6v00B72USpav6rDUUhqxAw)
    - Found and fixed several bugs related to data races in Eudico
    - Fixed Eudico's CI pipeline

### 🧵 Discussion 
- @jsoares: Where are we on finalising M2? Is this complete with [#68](https://github.com/filecoin-project/mir/pull/68)?
    - Yes, will close on merge
- Refine M3 goal

### 🎯 Up next
- @dnkolegov
    - Implementing new [Eudico Ordering Layer Interface](https://hackmd.io/6v00B72USpav6rDUUhqxAw)
- @sergefdrv 
    - Come up with testing strategy for Mir/ISS and start implementing it
- @matejpavlovic 
    - Refine timing estimates for M3 plan
    - Review SRDS papers

**Comment by matejpavlovic** [comment]
## 2022-05-30

### ✋ Attendees
- @adlrocha 
- @dnkolegov 
- @jsoares 
- @matejpavlovic
- @sergefdrv 
- @vukolic 

### 📣 Updates
- Denis:
    - Worked on Eudico ordering interface ([PR](https://github.com/filecoin-project/eudico/pull/202))
    - Expanded itests for Mir consensus: added tests with two nodes
    - Investigated how Eudico works
    - Updated the CI config to fix issues with Mir repo
    - Intriduced [SyncBlock](https://github.com/filecoin-project/eudico/pull/202/files#diff-feee0c0a03987b1f3b16b50c28138474398a2c93d381c808ec4c90d9dfcff844R109) funtion
- @sergefdrv 
    - Learning Filecoin spec
    - Draft of [Mir Framework Test Strategy](https://hackmd.io/OEIctHWCQV6tmJwVUGNX0Q)
    - Enabled [LGTM code analyzer for Mir](https://lgtm.com/projects/g/filecoin-project/mir/)
    - Added badges to Mir's ReadMe: [LGTM](https://github.com/filecoin-project/mir/pull/83), [GitHub CI](https://github.com/filecoin-project/mir/pull/84), [Go Reference](https://github.com/filecoin-project/mir/pull/85), [Go Report Card](https://github.com/filecoin-project/mir/pull/86)
- @matejpavlovic 
    - [Refined schedule for M3](https://hackmd.io/9SPHp_SWSHa8zbwXe9fO3Q)
    - [Proposal for Mir Generalization](https://github.com/filecoin-project/mir/pull/82)

### 🧵 Discussion 
- Addressing errors/warnings reported by code static analysis ([linter](https://github.com/filecoin-project/mir/runs/6653509881?check_suite_focus=true#step:4:25), [LGTM](https://lgtm.com/projects/g/filecoin-project/mir/alerts/), [GH Security](https://github.com/filecoin-project/mir/security/dependabot))
- @matejpavlovic : Dissemination of delivered blocks ([Denis' hack](https://github.com/filecoin-project/eudico/pull/202)) - @adlrocha is it viable?
- How does PoS relate to BFT? Do we implement any incentive mechanism?
  - PoS is a generalization of BFT, using different voting powers for different nodes.
  - The ordering layer gets this data from the outside (power table),
  no need to figure it out itself.
- How does HC work? Who creates subnets and how?
  - There is no succinct document describing it yet.
  - With the new FVM update, the filecoin will support creating subnets,
    each of them potentially using (an instance of) the BFT ordering layer.
  - The state of the subnet periodically writes its checkpoint to the parent.
- @matejpavlovic: Demo day - who shows what? Options:
    - @matejpavlovic shows demo chat app with a node crashing
    - @dnkolegov shows integration of mir into Eudico
        - with a node crashing (if possible)
        - without a node crashing

### 🎯 Up next
- @sergefdrv 
    - Discuss and finalize Mir testing strategy
    - Start improving testing of Mir
- @matejpavlovic
    - Polish BFT ordering interface
    - Pseudocode using the interface
    - Sync with @dnkolegov on demo and pick one option from above

**Comment by atonkikh** [comment]
Oh, hi! I'm just QA engineer from another company, I don't mind learning Rust, but can you help with DSL please.
Also, do you pay euros?

**Comment by xosmig** [comment]
@atonkikh oh, sorry, it's because I also sometimes use atonkikh as my nickname.
Don't worry, I'll handle the DSL thing :)

**Comment by matejpavlovic** [comment]
## 2022-06-27

### 📣 Updates
- @matejpavlovic
    - [Fix debug mode bug in Mir](https://github.com/filecoin-project/mir/pull/106) and a [debugger corner case](https://github.com/filecoin-project/mir/pull/110)
    - Code review of [Mir-Eudico crypto integration](https://github.com/filecoin-project/eudico/pull/208)
    - Completed [reconfiguration section](https://hackmd.io/P59lk4hnSBKN5ki5OblSFg?both#Re-Configuration)
      in the [ordering layer design document](https://hackmd.io/P59lk4hnSBKN5ki5OblSFg)
    - [Segment-level checkpoint in ISS](https://github.com/filecoin-project/mir/pull/113)
 - @sergefdrv 
     - Cleaning up existing Mir test infrastructure
         - [Enhance slow-propose test configuration](https://github.com/filecoin-project/mir/pull/109) 
         - [Implement restoring fake app state](https://github.com/filecoin-project/mir/pull/114)
         - [Replace Ginkgo and Gomega with Testify](https://github.com/filecoin-project/mir/pull/111)
     - Code review
         - [PBFT instance-level checkpoints](https://github.com/filecoin-project/mir/pull/113)
- @xosmig 
    - Merged [`ContextStore`](https://github.com/filecoin-project/mir/pull/103).
    - A lot of work on [DSL modules](https://github.com/filecoin-project/mir/pull/102):
        - Implement a user-friendly way to manage context using an internal `ContextStore`.
        - [Revisited the design](https://github.com/filecoin-project/mir/pull/102#issuecomment-1162382353) to avoid hacky pointer conversions.
        - Extensively [tested](https://github.com/filecoin-project/mir/blob/main/pkg/dsl/dslmodule_test.go) and documented.
        - **UPD:** merged [the PR](https://github.com/filecoin-project/mir/pull/102).
    - Working on the design draft for dynamically creating and deleting modules in Mir.
    - Started working on the availability layer.
- @dnkolegov:
    - [Implemented](https://github.com/filecoin-project/eudico/pull/208) Mir’s crypto interface
    - Working on the [new](https://github.com/filecoin-project/eudico/pull/214) ordering layer


### 🧵 Discussion 
- @xosmig: Do we already have some infrastructure to test individual modules?
  I am thinking how (and if) I should test the [BCB implementation](https://github.com/filecoin-project/mir/pull/102)
    - @matejpavlovic: To my best knowledge, we don’t have that yet.
      I don’t think testing the BCB implementation should be a priority now though.
      When Sergey is done with the new testing infrastructure, let’s see if and how that can be leveraged.
      @sergefdrv do you have anything in that direction in mind?
    - @sergefdrv: so far, I'm focused on integration testing, not module-level testing.
      I think one can write simple tests for individual modules as usual in Go. I would recommend using Testify.
    - @xosmig Yes, I did that for dsl modules (just calling `ApplyEvents` directly and checking the output), but BCB probably needs something more similar to integration tests: we need to run several nodes with other modules and the application being mocked, let the nodes talk to each other while manipulating the network, and check the outcome.

### 🎯 Up next
- @matejpavlovic:
    - Finish ISS protocol implementation (mostly message re-tramsmission)
    - Update and extend Eudico integration design documents
        - Specify details about (re-)configuration implementation
        - Update Eudico-Mir interface
- @sergefdrv 
    - Continue cleaning up existing Mir test infrastructure
    - Implement logical time in integration tests instead of using real timers and clocks. This will speed up testing and give 
- @xosmig 
    - Basic availability layer implementation.
    - Design draft for dynamic modules, without implementation.
- @dnkolegov
    - continue working on new Eudico ordering layer
    - implement more efficient FIFO mechanism in the Eudico block assembler
    - libp2p implementation for Mir

**Comment by matejpavlovic** [comment]
## 2022-07-04

Additional (finer) comments to this text can be found on [HackMD](https://hackmd.io/mYWvdfYvQgmG518tAZC0KQ)
### 📣 Updates
- @dnkolegov:
    - working on libp2p transport for Mir
- @xosmig:
    - working on a basic availability layer implementation
- @sergefdrv 
    - Another round of review on [PBFT instance-level checkpoints](https://github.com/filecoin-project/mir/pull/113)
- @matejpavlovic
    - Mostly non-Y3 work (HC review and BRAINS conference paper reviews)
    - Finalizing [PBFT instance-level checkpoints](https://github.com/filecoin-project/mir/pull/113)
    - Sorting out details transport layer and ordering layer implementation (for discussion summaries, see below).

### 🧵 Discussion 
- Libp2p transport for Mir:
    - @dnkolegov: @matejpavlovic  and I discussed today basic implementation of libp2p transport. Matej's suggestion is to use direct libp2p connections. In essense, that is similar to the current GRPC transport in Mir but with Libp2p. @adlrocha are you fine with that?
    - @matejpavlovic: In essence, the transport layer needs to provide authenticated point-to-point messaging. Using gossipsub is not desired for BFT-style consensus, since 1) not all messages need to be send to all nodes and 2) gossip latency is too high for most messages in BFT-style protocols.
    - @matejpavlovic: Aggregating from Comments: It seems mostly like a question of performance. Unless there is a really big performance difference, let's stick to one networking stack.
    - @matejpavlovic: I'd say we go for libp2p-based transport and once we have time at a later point, experiment with alternatives if we believe we can get a substantial performance gain.
- Details on [Eudico’s ordering layer implementation](https://hackmd.io/P59lk4hnSBKN5ki5OblSFg)
    - @xosmig: Last week, @matejpavlovic and I discussed the architecture of Eudico’s ordering Layer and came to a conclusion that, in order to avoid replicating the mempool structure in the availability layer module, the interface between the two should be pull-oriented (the availability layer asks for new blocks) rather than push-oriented (the mempool pushes new blocks/transactions to the availability layer). The main problem with the push-oriented interface was that the mempool could push new transactions at a faster rate than the availability layer could process them. Then we would need to include some buffers and non-trivial strategies for selecting and dropping transactions in the availability layer. On the other hand, pull-based architecture allows to encapsulate all such logic in the mempool module.
    - @xosmig: Also, we agreed to merge most of the functionality of batch store into the availability layer due to a high coupling between the functionalities of the two.
        - @matejpavlovic: Will update the design docs accordingly (see "Up Next" below)
    - @xosmig: Now I have a similar concern about the interface between the availability layer (composing and distributing TX batches for ordering, e.g. Narwhal) and the consensus layer: should the consensus layer also request (pull) blocks from the availability layer or should the availability layer push new blocks to the consensus layer?
        - There are some issues with both approaches:
            - Push-based: same as before, availability layer can go too fast or too slow.
            - Pull-based: implemented naively, may add extra latency.
                - It can be addressed if the availability layer works “one block in advance”. However, then there may be an issue with “recency” of the blocks (in case the consensus layer is rather slow).
        - @matejpavlovic:
            1. I think this can be addressed by making the Mir module that is emitting events representing the batches for ordering an `ActiveModule`. That way, if the consensus module is too slow to process them, the Mir implementation will temporarily stop reading new batches from the availability layer.
                - Comment discussion summary: This is a lower-level mechanism to protecting from going out of memory. Let's not use it for protocol-level flow control.
            3. This problem might actually happen to be prevented at a higher level. Given that the system works in epochs (between which it may reconfigure), it might be better to make even the availability layer only produce a finite number of batches per epoch. And since starting new epochs depends on the progress of the consensus protocol, we get an implicit flow control mechanism.
        - @xosmig: a few advantages of the pull-based model (with the optimization of working “one block in advance”):
            - lower coupling, cleaner interfaces (details like “epochs” don't “leak” into the availability layer implementation);
            - I think it may work nicer in combination with a more advanced availability layer like Narwhal because, in case of Narwhal, we probably do want it to go as fast as possible, but sometimes the consensus layer needs to get the latest tip of the dag and commit it (it would be able to do so through the request-response mechanism).
            - The problem of “recency” of blocks in case of slow consensus is less severe, albeit still present. (i.e., the delay between the creation of a block and the time it is proposed is smaller).
- A small question about Filecoin:
    - @xosmig: does the notion of a "valid transaction" depend on the state of the blockchain? Can / should we start working on a new block before deciding the previous one?
    - @matejpavlovic: While we could, in the future, support some "external validation" (informally discussed few times, last time during Consensus Factory), I would not go as far as requiring "closed-loop" block processing. For now (i.e. for the MVP) let's just consider everything that makes it in the mempool as valid (from the point of view of the ordering layer).
    - @xosmig ok, noted, thanks.
- Consensus proofs:
    - @sergefdrv: The HC spec mentiones: Subnets can run any consensus algorithm of their choosing, provided it can meet a defined interface, and they can determine the _consensus proofs_ they want to include for light clients. How do we plan to implement the consensus proofs in Mir/ISS?
    - @matejpavlovic: Currently we do not provide an explicit interface for "consensus proofs" (not sure whether and where one is defined for HC. @adlrocha?). But principally, the checkpoint certificates produced by the consensus implementation can serve as proofs, as they are signed by a quorum of the validators.
        - @sergefdrv: 
            - It is not quite clear what granularity is requered for the consensus proofs. One natural answer would be: per block, another: per subnet checkpoint. @adlrocha, what do you think?
            - @matejpavlovic, if it is per-block then how do we generate such proofs if we reassemble batched into block at the end of the ordering process?
        - @matejpavlovic: Indeed, if we wanted such a proof for every block, that could be inconvenient to implement, since Eudico blocks don't necessarily correspond to Mir batches.
            - @sergefdrv: Neither the HC checkpoints seem to correspond to the ISS checkpoints.
        - @xosmig: it is always possible to implement a lightweight external gadget that collects >2/3 signatures under a statement "the k-th block's hash is 0x...".
            - @sergefdrv: @xosmig, yep, I think it is easier to generate finality proofs outside, since doing so as a part of consensus implementation might be difficult (think about normal case vs. view changes etc.)

### 🎯 Up next
- @xosmig 
    - Basic availability layer implementation.
    - Design draft for dynamic modules, without implementation.
- @dnkolegov:
    - working on libp2p transport for Mir
- @matejpavlovic:
    - Update design documents based on latest discussions.
    - Define details and data structures for membership and reconfiguration.

**Comment by matejpavlovic** [comment]
## 2022-07-11

### 📣 Updates
- @dnkolegov:
    - working on libp2p transport for Mir
        - The [initial version](https://github.com/filecoin-project/mir/pull/126) is ready, it is very simple and written in the gRPC transport style
        - the gRPC transport will be used for testing purposes
- @sergefdrv 
    - Keep working on logical time in integration tests (in my dirty [wip](https://github.com/sergefdrv/mir/tree/wip-testsim) branch)
        - Sketched [logical time simulation interface](https://github.com/filecoin-project/mir/commit/195787fd34b795191990319d2d944101d7d1aad7#diff-a8364d0a3eeef60252f724042251a27ec311c9a2ae902c6e923d8498463dac1bR14-R159) that should work for Mir
        - Defined [abstract time implementation](https://github.com/filecoin-project/mir/commit/195787fd34b795191990319d2d944101d7d1aad7#diff-9c8b8b10ee25a09c1de5459f23864dcb391ab085c901897eb13225eca2fbe38cR16-R21) to use in timer module
- @matejpavlovic
    - Updated [Eudico ordering design document](https://hackmd.io/P59lk4hnSBKN5ki5OblSFg),
      added [new section on Merit (block rewards)](https://hackmd.io/P59lk4hnSBKN5ki5OblSFg#Merit-block-reward)
    - Finished [basic consensus implementation](https://github.com/filecoin-project/mir/issues/78)
        - [Instance-level checkpointing](https://github.com/filecoin-project/mir/pull/113)
        - [Checkpoint certificate verification](https://github.com/filecoin-project/mir/pull/123)
        - [Periodic re-sending of critical protocol messages](https://github.com/filecoin-project/mir/pull/122)
    - Updated [tracking GH issue](https://github.com/protocol/ConsensusLab/issues/63).
        - 2 sub-tasks marked closed (Interface draft and Consensus layer).
        - Mempool and Transport almost finished (by @dnkolegov), only need review and a bit of boilerplate around them.
        - Availability layer prototype (by @xosmig) implemented, only requires review and integration.
        - Testing being worked on by @sergefdrv.
        - Lef to tackle: Performance evaluation and Reconfiguration (State manager)
- @xosmig 
    - [Finished the implementation](https://github.com/filecoin-project/mir/pull/125) of the availability layer prototype.
    - [Summarized](https://github.com/filecoin-project/mir/discussions/127) design challenges and potential solutions for the dynamic modules in Mir.
    - [Removed hardcoded dependencies](https://github.com/filecoin-project/mir/pull/118) of the Mir core on ISS.
    - [Implemented a protoc plugin](https://github.com/filecoin-project/mir/pull/121) to generate more generic-friendly code and simplified the dsl module implementation using it.
    - In parallel, exploring different directions to further improve Mir:
        - [Playing around with the protoc plugin](https://github.com/xosmig/mir/tree/boilerplate-plugin-main) to maybe generate all the boilerplate associated with programming on Mir.
        - Working on the concept of “submodules” as a simpler alternative to “dynamic modules”.

### 🧵 Discussion 
- @dnkolegov: @adlrocha could you review Mir's libp2p transport or recommend someone from the libp2p team for that?
    - @dnkolegov to ping @adlrocha when there is code to review.
- @matejpavlovic: [A few questions on the testing framework](https://github.com/filecoin-project/mir/discussions/128). Mostly for @sergefdrv.

### 🎯 Up next
- @dnkolegov:
    - integration Eudico with Mir's libp2p transport
    - adopting libp2p transport in the Mir's sample app
    - will discuss the next task within the Mir MVP project with @matejpavlovic until this Friday
- @sergefdrv 
    - Finish implementing logical time in integration tests
- @matejpavlovic
    - Refine Y3 roadmap to be similar type as B3
    - Refine Reconfiguration and Performance issues to more concrete AIs
    - Execute some of above AIs
- @xosmig 
    - Merge the availability layer prototype PR.
    - Add some integration tests for the availability layer.
    - Start adding complexity to the availability layer (message retransmission, persistent storage, etc.)

**Comment by matejpavlovic** [comment]
## 2022-07-25

### 📣 Updates
- dnkolegov:
    - [finished](https://github.com/filecoin-project/eudico/pull/223) integration of Mir with libp2p transport into Eudico
    - added several small PRs to clean Mir and improve integration with Eudico: [1](https://github.com/filecoin-project/mir/pull/150), [2](https://github.com/filecoin-project/mir/pull/149), [3](https://github.com/filecoin-project/mir/pull/142)
    - experimented with orchestration and reconfiguration: the main result is the question in the Discussion section
    - reviewed several PRs
    - I am still not sure about the correctness of the libp2p transport in Mir (opening connections). I asked the [question](https://filecoinproject.slack.com/archives/C03K82MU486/p1657869778870039) on libp2p-implementers channel. Nobody has responded yet.
- @sergefdrv 
    - Draft implementation of [Reproducible integration testing with simulated time](https://github.com/filecoin-project/mir/pull/148) (generally ready for review, but requires rebasing)
- @matejpavlovic
    - OOO
- @xosmig 
    - [Basic module mocking in Mir](https://github.com/filecoin-project/mir/pull/141)
    - [Placeholders library](https://github.com/xosmig/placeholders) for more advanced mocking capabilities
    - Finished refactoring the testing infrastructure: [PR1](https://github.com/filecoin-project/mir/pull/143), [PR2](https://github.com/filecoin-project/mir/pull/147)
    - [Implemented a prototype](https://github.com/filecoin-project/mir/pull/144) of the mempool module

### 🧵 Discussion 
- dnkolegov: 
    - @matejpavlovic suggested to create a new Mir node on each reconfiguration via Eudico's manager (to start with something and to have something for MVP). I experimented with this idea and found that this works not well because we have to stop the network transport and then reestablish all network connections. I was not able to make Mir nodes with libp2p transport work immediately after stopping the node, but it worked with adding Sleep command. I would suggest to improve this idea a little bit in the following way: let's add new connections into network transport and remove the old connections if necessary instead of calling `net.Stop()`, `net.Start()` and `net.Connect()` on each reconfiguration. So we need to add a function like NewConnections(map[NodeID]NetAddr) to the network transport interface. This function opens connections to nodes contained in the map without removing all network connection from previous stage.
        - @matejpavlovic: Yes I agree, let's re-use the existing connections.
          Conclusion: @dnkolegov to implement it and use it in the Eudico integration code.

### 🎯 Up next
- dnkolegov: 
    - discuss with @matejpavlovic the next tasks (I may be working on Mir reconfiguration support in Eudico)
    - address issue [224](https://github.com/filecoin-project/eudico/issues/224): I will find the latest commit where Tendermint works and push it in a separate branch on Eudico
- @sergefdrv 
    - Rebase [Reproducible integration testing with simulated time](https://github.com/filecoin-project/mir/pull/148)
    - Contribute to Eudico’s State Manager design and implementation
- @matejpavlovic
    - Finalize new roadmap for Y3-M3
    - Start contribute to reconfiguration-related code
- @xosmig 
    - OOO

**Comment by matejpavlovic** [comment]
## 2022-08-01

### 📣 Updates
- @sergefdrv 
    - Rebased and adjusted implementation of [Reproducible integration testing with simulated time](https://github.com/filecoin-project/mir/pull/148) (merged)
    - Made [CI run integration tests with simulated time](https://github.com/filecoin-project/mir/pull/161) (no new tests yet)
- dnkolegov:
    - Modified [Mir transport interface](https://github.com/filecoin-project/mir/pull/155)
    - Minor improvements and bug fixes ([1](https://github.com/filecoin-project/mir/pull/156), [2](https://github.com/filecoin-project/mir/pull/159))
    - [Refactored](https://github.com/filecoin-project/eudico/pull/234) Eudico code to support new transport interface and reconfiguration mechanics
- @matejpavlovic
    - Updated [Y3-M3 Roadmap](https://hackmd.io/e4nU9jmGSJe4FmNSpHIjCg). Slipping a bit, the new realistic delivery M3 delivery date including membership reconfiguration is now Aug 26th. Other features should still be done this week.
    - [Outline](https://hackmd.io/e4nU9jmGSJe4FmNSpHIjCg#Node-joining) of reconfiguration implementation
    - [More](https://github.com/filecoin-project/mir/pull/152) [Mir](https://github.com/filecoin-project/mir/pull/154) [improvements](https://github.com/filecoin-project/mir/pull/160)
    - Started working on node joining in the chat demo app
- @xosmig 
    - Was OOO

### 🧵 Discussion 
- The outcome of most discussions is captured in the new [Y3-M3 Roadmap document](https://hackmd.io/e4nU9jmGSJe4FmNSpHIjCg).
- Discussions on demo day presentations in the [CL Meeting Notes](https://docs.google.com/document/d/1eHJxNHkZ8D9WqHDt7FhPdwaK6QGrQx9hV1vAintBuyQ/edit#heading=h.4296cp8skzjc).

### 🎯 Up next
- dnkolegov:
    - Implement membership update submission
    - Integrate the Mir availability layer into Eudico
- @sergefdrv 
    - Implement Load generator for Mir-based orderer
- @matejpavlovic
    - Integrate availability layer in chat app
    - If time permits, continue making chat app reconfigurable
- @xosmig 
    - Dsl modules Demo at [MoaDD](https://docs.google.com/document/d/1tk-SuFFUpglImoTIEdGAku3qFsoE5a47G0u3VJO8T_k)

**Comment by matejpavlovic** [comment]
## 2022-08-08

### 📣 Updates
- @dnkolegov:
    - spent a lot of time trying to find the reason of [this](https://github.com/filecoin-project/eudico/issues/235) data race and why we have not seen that before
    - draft implementation of the Mir reconfiguration in Eudico
    - drafted itests for reconfiguration
- @xosmig:
    - [MoaDD](https://drive.google.com/drive/folders/1m8ys4_u-LqeTLOtUtNZYr7Pt_cUmFlGH) presentation
    - A bunch of small code improvements: [link1](https://github.com/filecoin-project/mir/pull/164), [link2](https://github.com/filecoin-project/mir/pull/166), [link3](https://github.com/filecoin-project/mir/pull/167), [link4](https://github.com/filecoin-project/mir/pull/168), [link5](https://github.com/filecoin-project/mir/pull/169), [link6](https://github.com/filecoin-project/mir/pull/170), [link7](https://github.com/filecoin-project/mir/pull/172), [link8](https://github.com/filecoin-project/mir/pull/175)
    - [Consistent Broadcast](https://github.com/filecoin-project/mir/pull/176) and [Availability Layer](https://github.com/filecoin-project/mir/pull/171) demos
- @matejpavlovic
    - Initial (hacky) implementation of [reconfiguration of the chat demo app](https://github.com/matejpavlovic/mir/tree/chat-reconfig-clean). Currently, the newly joining node needs to get the information about the configuration out-of-band, but with that information it can join the network. For the chat demo app, this is necessary. For Eudico, the idea is to have it obtain the information from the parent network.
### 🧵 Discussion 
- @dnkolegov: @matejpavlovic @xosmig, is the availability layer is ready? do (will) we  have an example (in the chat-app) of how to use it properly?
    - @xosmig: There is a separate demo for the availability layer: [link](https://github.com/filecoin-project/mir/pull/171). Afaik, it is not yet integrated into ISS (and, hence, to the chat-demo).
    - @matejpavlovic I switched up the chat demo reconfiguration and the availability layer priorities, since the reconfiguration seemed more pressing wrt. Denis' work. I'll focus on the availability today and tomorrow, so it's integrated with ISS by tomorrow evening.
- @xosmig: I am a bit concerned with the current approach to technical debt in the code. We should probably aim at writing a bit more general future-proof code. Otherwise, it is just being rewritten twice. In most cases, doing things general enough the first time is really not a problem. A few examples:
    - My personal big strugle: Mir != ISS (as I am tasked to do first non-ISS protocols in Mir). Unfortunatelly, I keep seeing (unnecessarily) ISS-specific code being added regularly.
    - `strncov.Atoi(nodeID)`
    - `if tr.Sim != nil {...}`
    - `if transport, ok := tr.Modules["net"]; ok {...}` (guilty, this one is mine)
    - @matejpavlovic: I acknowledge this point and agree we should do better on this. To explain how we got there initially: Historically, Mir started out as an implementation of ISS and has progressively been generalized to be less and less ISS-specific. But admittedly, few parts of the code that are ISS-specific are still from the old times. Some parts were just written in an ISS-specific way because it was cognitively simpler at that time, and generalization is lagging (leaving it as debt). I agree this is sub-optimal.
        - My suggestion: Let us be stricter with this from now on and, when reviewing PRs, point out ISS-specific parts and request generalization. And ingeneral, push for only merging PRs that do not add more debt, and in rare cases where they do, it should be justified and agreed-upon.
    - @xosmig: I understand why the code was written this way initially. I am more concerned about new code being added "by analogy" with the old code, thus increasing technical debt while we should aim at gradyally paying it off.
### 🎯 Up next
- dnkolegov:
    - continue to work on Mir reconfuguration in Eudico
- @matejpavlovic
    - Availability
    - Reconfiguration (node joining and state transfer)
- @xosmig
    - Hopefully finally start implementing Narwhal

**Comment by matejpavlovic** [comment]
## 2022-08-15

### 📣 Updates
- @dnkolegov:
    - I am still working on Mir [reconfiguration](https://github.com/filecoin-project/eudico/pull/234) in Eudico
        - wrote tests
        - fixed bugs in old tests
        - fixed docs
        - prepared a test/demo scenario
    - PR review - [1](https://github.com/filecoin-project/mir/pull/185), [2](https://github.com/filecoin-project/mir/pull/184), [3](https://github.com/filecoin-project/mir/pull/180)
- @xosmig:
    - Working on Narwhal impl in Go
    - [Improving](https://github.com/filecoin-project/mir/pull/188) Mir
    - [Experimenting](https://github.com/xosmig/mir/pull/2) with improving DSL modules
- @matejpavlovic
    - [Mir now supports reconfiguration](https://github.com/filecoin-project/mir/pull/180). Node joining is tested with the chat app. Good candidate for the next Demo Day. Not secure yet due to partly stubby certificate verification code.
    - [Separate availability (mempool, content dissemination) layer integration with ISS](https://github.com/filecoin-project/mir/pull/183)
    
### 🧵 Discussion 
- Summary of sync brainstorming between @matejpavlovic and @xosmig about Mir, Narwhal, and reconfiguration:
    - Narwhal implementation:
        - What persistent storage system to use? BadgerDB was used just as an example in the design document, because it happens to be used by the (inherited) WAL implementation. In principle, @matejpavlovic has no concrete preference for a particular data store. What potentially might be relevant is [go-datastore](https://github.com/ipfs/go-datastore).
            - [Rust implementation](https://github.com/facebookresearch/narwhal) uses [RocksDB](http://rocksdb.org/).
        - The implementation will rely on a reliable communication abstraction. As for now, the unreliable Net module can be temporarily used as a stub. @matejpavlovic to check out the updated [GitHub discussion on reliable message delivery](https://github.com/filecoin-project/mir/discussions/132).
    - Narwhal garbage collection:
        - The garbage collection in Narwhal "cuts off" some dependencies. It may become a problem when we will be adding deduplication and censorship resistence. It is not immediately clear whether it's actually necessary: maybe we could garbage-collect only ordered blocks and add "weak references" as in DagRider to make sure every block is eventually ordered? Let's look more closely at the DagRider and think more about it.
    - Narwhal reconfiguration:
        - Multiple instances of Narwhal will co-exist, using a similar instantiation and garbage collection mechanism as ISS. The ordering layer (ISS in our case) already drives the advancing of epochs and configurations and the creation of checkpoints. It will simply emit analogous events to the availability module, which will instantiate and garbage-collect Narwhal instances accordingly.
        - As the ordering layer (ISS) is pulling its input from the availability layer (Narwhal), the availability certificates need not carry explicit epoch information. The ordering layer will simply pull the certificate from the corresponding instance of Narwhal. 
        - The instances of Narwhal will be represented as nested modules within a main availability module.
    - Mir nested modules:
        - In order to be able to kill old instances of the availability layer module and create new ones during the reconfiguration, we need something like [dynamic modules](https://github.com/filecoin-project/mir/discussions/127). However, we don't have a clean design for dynamic modules yet. Instead, we settled on a simpler alternative, which can be implemented with a tiny modification to how Mir core processes module ids.
            - TODO(@xosmig): write down the design for submodules either in discussions or in an ADR.
            - TODO(@matejpavlovic): Introduce structured module identifiers to enable suport for nested modules

### 🎯 Up next
- @dnkolegov:
    - finish integration of Mir reconfiguration
    - start working in integration of Mir's availability layer into Eudico
- @xosmig:
    - A simple stubby Narwhal prototype.
- @matejpavlovic
    - Merge reconfiguration and availability layer support
    - Prepare chat reconfiguration demo

**Comment by matejpavlovic** [comment]
## 2022-08-22

### 📣 Updates
- @dnkolegov:
    - [Finished](https://github.com/filecoin-project/eudico/pull/234) adding part of Mir reconfiguration in Eudico
    - [Improved](https://github.com/filecoin-project/mir/pull/190) Mir libp2p transport
- @xosmig:
    - Some work on the availability layer: [PR](https://github.com/filecoin-project/mir/pull/189).
    - Couldn't resist anymore and decided to implement code generation (in [WIP branch](https://github.com/xosmig/mir/tree/code-gen-alt)). It turned out more difficult than I expected, but it is almost finished at this point.
- @matejpavlovic
    - The availability layer (multisig collector) is now also [reconfigurable using "factory modules"](https://github.com/filecoin-project/mir/pull/192).
    - [Wrote](https://github.com/filecoin-project/mir/pull/182)
      [lots](https://github.com/filecoin-project/mir/pull/191)
      [of](https://github.com/filecoin-project/mir/pull/193)
      [other](https://github.com/filecoin-project/mir/pull/194)
      [code](https://github.com/filecoin-project/mir/pull/195)
      for Mir.

### 🧵 Discussion 
- @xosmig: What do we want from code generation? 
    - @xosmig: For example, I can generate a hierarchy of structs completely separate from the ones generated by protobufs, with Mir types (e.g., `t.NodeID` instead of `string`) and conversion functions.
        - @matejpavlovic: That would be nice to have. 
            - Would that be generated from the `.proto` definitions?
                - @xosmig: the generator uses the protoc-generated `.pb.go` files as input.
                  This choice was motivated by relative simplicity compared to working directly with `.proto` 
                  files since protobufs have their own type system which would need to be translated to go type system.
            - How would the generator know the difference between two different Mir types that map to the same protbuf type? 
              E.g., if the protobuf message defines a field of type uint64, how would the generator know that it is, say, 
              a sequence number and not an epoch number? Is it possible add additional annotations in the `.proto` file?
                - @xosmig: the generator looks at optional annotations on fields in `.proto` files, e.g.: 
                  `string module = 1 [(mir.type) = "github.com/filecoin-project/mir/pkg/types.ModuleID"];`
            - It would also be great to have, for each generated type, a generated constructor function that takes the fields as arguments.
              I would then encourage a convention of never creating an object directly through `SomeObjectType{Field: val, ...}`
              and always using the constructor. This way, should we decide to add fields to some objects,
              the compiler would help avoiding the default zero values for that field in instantiations of the object
              that one might forget to update.
                - @xosmig: yes, I already implemented the constructor functions.
    - @xosmig: The baseline is that I will just generate constructors and DSL functions.
        - @matejpavlovic: Yes that's a very helpful thing. Does this also mean constructors for event protobufs?
            - I have the code to generate the constructors implemented. We need to decide to which extent we want to use it.
            - @matejpavlovic: I'd say there is no harm in generating a constructor for each event type, even if it ends up not being used.
        - @matejpavlovic: A skeleton implementation of the DSL module would also be nice, with generated handlers for each event type.
            - For this we would need to use something like `service` description in protobufs. 
              This would be nice in the future, but I think goes beyond the intended scope for the first version of code generator.
            - @matejpavlovic: A coarse but simple (and I agree, not necessarily very pretty) way would be
              to just generate an handler for every single event type defined for the package (even for those meant as output).
              The user would then delete those they don't use.
              Although, that might require the user to edit the generated intermediate "event registration" files.
              I guess that could be maybe solved by another annotation, but I agree that this is way out of scope for now.
- @matejpavlovic: Also in the light of the above, and as @xosmig already suggested at some point,
  we should clean up the protobuf definitions, as now they are quite messy.
  The `events.proto` and `messages.proto` are historical artifacts and a big part (if not all) of their content
  should be moved to files associated with the respective modules. I'll soon have a proper look at that and move things around.
  @xosmig if this has any impact on the code generation and you would like me to stick to some particular conventions, pls let me know.
  - @xosmig: I think there is an implicit convention for event and message subtypes 
    (e.g., `isspb.Event`, `availabilitypb.Event`, `isspb.ISSMessage`, `mscpb.Message`).
    - @matejpavlovic: Yes that makes sense. (The `isspb.ISSMessage` should then be renamed to `isspb.Message` though.)
  - I also don't want to support `oneof`s with primitive type options.
    Partly, because I am considering a posibility of the future where we will move to a different data model with interfaces instead of oneofs.
    - Ok, note taken, that makes sense too. I won't use primitive `oneof`s and when refactoring, I'll remove the existing ones.

### 🎯 Up next
- @dnkolegov:
    - Implement reconfiguration by removing Mir nodes
    - Integrate the Mir availability layer into Eudico 
- @xosmig:
    - Finish the code generation
    - Narwhal prototype
- @matejpavlovic
    - Polish reconfiguration
    - Start performance measurements

**Comment by matejpavlovic** [comment]
## 2022-09-19

### 📣 Updates
- @matejpavlovic
    - Prioritized [roadmap for next milestone](https://hackmd.io/dBW9nJrdRXKrC4wq0iJOHA)
    - Mir improvements:
        - [SMR system abstraction](https://github.com/filecoin-project/mir/pull/224)
        - Several technical improvemens: [1](https://github.com/filecoin-project/mir/pull/201) [2](https://github.com/filecoin-project/mir/pull/205) [3](https://github.com/filecoin-project/mir/pull/214) [4](https://github.com/filecoin-project/mir/pull/223) [5](https://github.com/filecoin-project/mir/pull/225)
    - Sergey switching to IPC and Rust-based projects. (Thanks @sergefdrv for your contribuons!)
    - First external open-source contribution! :tada:
        - [Threshold crypto](https://github.com/filecoin-project/mir/pull/219)
    - Project proposals (for interns, new collaborators, anyone...):
        - [Mir Diagnostics and Fault Injection (`mircat`)](https://github.com/filecoin-project/mir/issues/220)
        - [Mir-based SMR Performance](https://github.com/filecoin-project/mir/issues/221)
        - [Crash recovery](https://github.com/filecoin-project/mir/issues/222)
- @xosmig 
    - Was busy with research related stuff and preparing a final version for a ConsensusDay paper.
- @sergefdrv 
    - [Updated](https://github.com/filecoin-project/mir/pull/177#commits-pushed-c1610c1) benchmarking tools to use SMR abstraction
- @dnkolegov:
    - worked on the Mir [mempool](https://github.com/filecoin-project/eudico/pull/253) adapted to Eudico
    - tried to add semgrep linters to the Mir project but reverted, because implementation security is not top priority at present

### 🎯 Up next
- @matejpavlovic
    - Y3-M4 roadmap timing and assignment of DRIs
    - Sync roadmaps with BuilderNet
    - Mir preliminary perf evaluation
- @sergefdrv 
    - Update [benchmarking tools](https://github.com/filecoin-project/mir/pull/177) to use the new “SMR system” abstraction
- @xosmig 
    - Finish the code generation
- @dnkolegov:
    - Reestablishing lost network connections
    - implementing client watermark windows for limiting the number of in-flight transactions

**Comment by matejpavlovic** [comment]
Y3 has been merged with [the B4 project](https://github.com/protocol/ConsensusLab/issues/7), including the [meeting notes](/tdItavAETRaUbETbp3pLlA).
